Branch-SDK/src/io/branch/referral/indexing/BranchUniversalObject.java

```diff
@@ -0,0 +1,506 @@
+    /* Any keyword associated with the content. Used for indexing */
+    private final ArrayList<String> keyWords_;
+    /* Expiry date for the content and any associated links. Represented as epoch milli second */
+    private long expiration_;
+
+    private LinkProperties linkProperties_;
+
+    /**
+     * Defines the Content indexing modes
+     * PUBLIC | PRIVATE
+     */
+    public enum CONTENT_INDEX_MODE {
+        PUBLIC, /* Referred contents are publically indexable */
+        // PUBLIC_IN_APP, /* Referred contents are publically available to the any app user */
+        PRIVATE/* Referred contents are not publically indexable */
+    }
+
+
+    /**
+     * <p>
+     * Create a BranchUniversalObject with the given content.
+     * </p>
+     */
+    public BranchUniversalObject() {
+        metadata_ = new HashMap<>();
+        keyWords_ = new ArrayList<>();
+        canonicalIdentifier_ = "";
+        title_ = "";
+        description_ = "";
+        type_ = "";
+        indexMode_ = CONTENT_INDEX_MODE.PUBLIC; // Default content indexing mode is public
+        expiration_ = 0L;
+        linkProperties_ = new LinkProperties();
+    }
+
+    /**
+     * <p>
+     * Set the canonical identifier for this BranchUniversalObject. Canonical identifier is normally the canonical path for your content in the application or web
+     * </p>
+     *
+     * @param canonicalIdentifier A {@link String} with value for the canonical identifier
+     * @return This instance to allow for chaining of calls to set methods
+     */
+    public BranchUniversalObject setCanonicalIdentifier(String canonicalIdentifier) {
+        this.canonicalIdentifier_ = canonicalIdentifier;
+        return this;
+    }
+
+    /**
+     * <p>
+     * Set a title for the content referred by this object
+     * </p>
+     *
+     * @param title A {@link String} with value of for the content title
+     * @return This instance to allow for chaining of calls to set methods
+     */
+    public BranchUniversalObject setTitle(String title) {
+        this.title_ = title;
+        return this;
+    }
+
+    /**
+     * <p/>
+     * Set description for the content for the content referred by this object
+     * <p/>
+     *
+     * @param description A {@link String} with value for the description of the content referred by this object
+     * @return This instance to allow for chaining of calls to set methods
+     */
+    public BranchUniversalObject setContentDescription(String description) {
+        this.description_ = description;
+        return this;
+    }
+
+    /**
+     * <p/>
+     * Set the url to any image associated with this content.
+     * <p/>
+     *
+     * @param imageUrl A {@link String} specifying a url to an image associated with content referred by this object
+     * @return This instance to allow for chaining of calls to set methods
+     */
+    public BranchUniversalObject setContentImageUrl(String imageUrl) {
```

Comment:
sorry to nitpick about this now: @aaustin are we okay having methods called "set_Content_ImageUrl" when we moved away from calling this a piece of content to a universal object? `setContentDescription` makes sense if `setDescription` would collide with OS-provided methods (on iOS, there is already something called a description)